pkg/envoy/ads/response.go

```diff
@@ -18,74 +19,80 @@ const (
-
-	discoveryResponse, err := s.newAggregatedDiscoveryResponse(proxy, req, cfg)
-	if err != nil {
-		log.Error().Err(err).Msgf("[%s] Failed to create response for proxy with SerialNumber=%s on Pod with UID=%s", xdsShortName, proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
+	if discoveryResponse, err := s.newAggregatedDiscoveryResponse(proxy, req, cfg); err != nil {
+		log.Error().Err(err).Msgf("[%s] Failed to create response for proxy with SerialNumber=%s on Pod with UID=%s", typeURI.Short(), proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
+		xdsPathTimeTrack(startedAt, log.Debug(), typeURI, proxy, false)
 		return err
-	}
-
-	if err := (*server).Send(discoveryResponse); err != nil {
-		log.Error().Err(err).Msgf("[%s] Error sending to proxy with SerialNumber=%s on Pod with UID=%s", xdsShortName, proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
+	} else if err := (*server).Send(discoveryResponse); err != nil {
+		log.Error().Err(err).Msgf("[%s] Error sending to proxy with SerialNumber=%s on Pod with UID=%s", typeURI.Short(), proxy.GetCertificateSerialNumber(), proxy.GetPodUID())
+		xdsPathTimeTrack(startedAt, log.Debug(), typeURI, proxy, false)
 		return err
 	}
 
-	success = true // read by deferred function
+	xdsPathTimeTrack(startedAt, log.Debug(), typeURI, proxy, true)
 	return nil
 }
 
 // sendResponse takes a set of TypeURIs which will be called to generate the xDS resources
 // for, and will have them sent to the proxy server.
 // If no DiscoveryRequest is passed, an empty one for the TypeURI is created
+// TODO(draychev): Convert to variadic function: https://github.com/openservicemesh/osm/issues/3127
 func (s *Server) sendResponse(typeURIsToSend mapset.Set,
 	proxy *envoy.Proxy,
 	server *xds_discovery.AggregatedDiscoveryService_StreamAggregatedResourcesServer,
 	request *xds_discovery.DiscoveryRequest,
 	cfg configurator.Configurator) error {
-	success := true
-	if typeURIsToSend.Cardinality() == len(envoy.XDSResponseOrder) {
-		defer xdsPathTimeTrack(time.Now(), log.Info(), ADSUpdateStr, proxy.GetCertificateSerialNumber().String(), &success)
-	}
-
 	// Order is important: CDS, EDS, LDS, RDS
 	// See: https://github.com/envoyproxy/go-control-plane/issues/59
+	var typesToSend []envoy.TypeURI
 	for _, typeURI := range envoy.XDSResponseOrder {
-		if !typeURIsToSend.Contains(typeURI) {
-			continue
+		if typeURIsToSend.Contains(typeURI) {
+			typesToSend = append(typesToSend, typeURI)
 		}
+	}
+
+	if len(typesToSend) < 1 {
+		// This should never happen.
+		log.Error().Msg("The list of discovery responses passed to sendResponse() is empty. This is an invalid invocation of sendResponse().")
+		return errors.New("list of discovery responses is empty")
+	}
+
+	thereWereErrors := false
+
+	// Order is important: CDS, EDS, LDS, RDS
+	// See: https://github.com/envoyproxy/go-control-plane/issues/59
+	for _, typeURI := range typesToSend {
+		fullUpdateRequested := request != nil && envoy.TypeURI(request.TypeUrl).IsWildcard()
```

Comment:
This prevents the use of wildcard with no request. Not arguing the implications, as it is not currently used (nor good or bad), but just stating to make sure this is a change from the logic on the left.

Reply:
I see. It seems that this should handle `request == nil`?
Would it make sense for this to be 		
```go
fullUpdateRequested := request == nil || (request != nil && envoy.TypeURI(request.TypeUrl).IsWildcard())
```

Reply:
I noticed that a few lines below we already had:
```go
if request == nil || fullUpdateRequested {
```

Which should be combined with:
```go
fullUpdateRequested := request != nil && envoy.TypeURI(request.TypeUrl).IsWildcard()
```

and the result is simply:
```go
fullUpdateRequested := request == nil || envoy.TypeURI(request.TypeUrl).IsWildcard()
```

I think this reads better and is correct, right?

Reply:
yep, that was something I forgot to comment on. This looks better IMO